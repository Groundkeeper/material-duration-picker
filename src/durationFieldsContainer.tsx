import {Grid, GridProps} from '@material-ui/core';
import * as React from 'react'
import {Dispatch} from 'react'
import {DurationField, DurationFieldProps} from "./durationField";
import {DurationType, DurationView, Labels} from "./types";
import DefaultLabels from "./defaultLabelsEn.json";
import {getDurationOverflow} from "./utils";


export type DurationFieldsContainerProps = {
  views: DurationView[];
  duration: DurationType;
  setDuration: Dispatch<DurationType>
  labels?: Labels

  GridContainerProps?: GridProps;
  GridItemProps?: GridProps;
  DurationFieldProps?: Partial<DurationFieldProps>
}

export const DurationFieldsContainer = ({
  views = [],
  duration,
  setDuration,
  labels: _labels,
  GridItemProps,
  GridContainerProps,
  DurationFieldProps
}: DurationFieldsContainerProps)=>{
  const labels = {
    ...DefaultLabels,
    ..._labels
  }

  return (
    <Grid container spacing={2} justify='space-around' {...GridContainerProps}>
      {views.map((view, i) => {
        const acc = i === 0 ? getDurationOverflow(duration, view) : 0

        const value = ((duration[view] || 0) + acc) || null

        return (
          <Grid item key={i} {...GridItemProps}>
            <DurationField
              label={labels[view]}
              value={value}
              onConfirm={(v) => {
                setDuration({
                  ...duration,
                  [view]: (v || 0) - acc
                })
              }}
              {...DurationFieldProps}
            />
          </Grid>
        );
      })}
    </Grid>
  );
}

