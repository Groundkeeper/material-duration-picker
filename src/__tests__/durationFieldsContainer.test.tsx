import React from 'react'
import {render, waitFor, screen, fireEvent} from '@testing-library/react'
import {DurationFieldsContainer, DurationFieldsContainerProps} from "../durationFieldsContainer";

describe('DurationFieldsContainer', () => {
  const setDurationMock = jest.fn()
  const emptyDuration = {};

  const Comp = ({
    views = ['minutes', 'seconds'],
    duration = {...emptyDuration, minutes: 5, seconds: 2},
    setDuration = setDurationMock,
    ...props
  }: Partial<DurationFieldsContainerProps>) => {
    return (
      <DurationFieldsContainer
        {...{
          views,
          duration,
          setDuration,
          DurationFieldProps: {
            inputProps: {
              title: 'fieldTitle',
            },
            ...props.DurationFieldProps
          },
          ...props
        }}
      />
    )
  };

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(
        render(
          <Comp/>
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  it('can render less views with greater duration', async () => {
    const utils = render(
      <Comp duration={{
        ...emptyDuration,
        hours: 1,
        minutes: 1,
        seconds: 1
      }}/>
    );
    const fields: any[] = utils.getAllByTitle('fieldTitle')
    expect(fields[0].value).toBe("61")
    expect(fields[1].value).toBe("1")
  });

  it('can setDuration to greater than views show', async () => {
    const utils = render(
      <Comp duration={{
        ...emptyDuration,
        minutes: 1,
        seconds: 1
      }}/>
    );

    const fields: any[] = utils.getAllByTitle('fieldTitle')

    fireEvent.change(fields[0], {target: {value: '61'}})
    fireEvent.blur(fields[0])

    expect(fields[0].value).toBe('61')

    expect(setDurationMock).toHaveBeenCalledWith({
      weeks: undefined,
      days: undefined,
      hours: undefined,
      minutes: 61,
      seconds: 1
    })
  })
})
