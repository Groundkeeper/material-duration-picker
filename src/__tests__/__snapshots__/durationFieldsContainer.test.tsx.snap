// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationFieldsContainer can match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-space-around"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            title="fieldTitle"
            type="number"
            value="5"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            title="fieldTitle"
            type="number"
            value="2"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
